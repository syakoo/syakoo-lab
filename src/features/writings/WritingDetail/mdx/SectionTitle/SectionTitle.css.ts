import { style } from "@vanilla-extract/css";

import { tokens } from "@/design-system/tokens";

export const sectionTitleStyle = style({
  marginTop: tokens.spaces[500],
  marginBottom: tokens.spaces[200],
  paddingBlock: tokens.spaces[200],
  fontSize: tokens.fontSizes[500],
  scrollMarginTop: tokens.spaces[200],
  position: "relative",

  selectors: {
    "&:before": {
      position: "absolute",
      top: "0",
      left: `-${tokens.spaces[200]}`,
      display: "inline-block",
      width: "1em",
      height: "4px",
      content: "",
      background: `linear-gradient(to right bottom, ${tokens.colors.brand.primary}, ${tokens.colors.brand.secondary})`,
      borderTopRightRadius: tokens.radii[100],
    },
  },
});

export const subsectionTitleStyle = style({
  marginTop: tokens.spaces[400],
  marginBottom: tokens.spaces[200],
  fontSize: tokens.fontSizes[300],
  scrollMarginTop: tokens.spaces[200],
});
