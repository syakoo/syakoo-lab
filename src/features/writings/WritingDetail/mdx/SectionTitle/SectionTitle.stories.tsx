import type { Meta } from "@storybook/react";

import { SectionTitle, SubSectionTitle } from ".";

const meta = {
  component: SectionTitle,
  tags: ["autodocs"],
  parameters: { testLevel: "snapshot" },
} satisfies Meta<typeof SectionTitle>;

export default meta;

export const Sample = () => <SectionTitle>Sample Title</SectionTitle>;

export const ContainElements = () => (
  <SectionTitle>
    Sample <code>Title</code>
  </SectionTitle>
);

export const SubSection = () => (
  <SubSectionTitle>Sample SubTitle</SubSectionTitle>
);
