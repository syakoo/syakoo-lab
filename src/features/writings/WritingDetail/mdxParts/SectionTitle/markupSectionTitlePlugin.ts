import { JSDOM } from "jsdom";
import React from "react";

import { resolveMDXAsComponent } from "@/features/mdx/resolveMDXAsComponent";
import { serializeMDX } from "@/features/mdx/serializeMDX";
import { MDXCustomTextPlugin } from "@/features/mdx/types";

/**
 * 見出し2, 3 (`## xxx`, `### xxx`) を Section, SubSection に変換するプラグイン
 */
export const markupSectionTitlePlugin: MDXCustomTextPlugin = async (mdText) => {
  const { renderToString } = await import("react-dom/server");
  const splittedMdTexts = mdText.split("\n");

  // カウントの仕組み
  const counter = {
    data: {} as Record<string, number>,
    count: function (key: string) {
      if (key in this.data) {
        this.data[key] += 1;
      } else {
        this.data[key] = 0;
      }

      return this.data[key];
    },
  };

  const resultSplittedTexts = await Promise.all(
    splittedMdTexts.map(async (text) => {
      /**
       * 描画しテキスト部分を取得する関数
       */
      const getInnerTextContent = async (content: string) => {
        // HACK: 見出しが無限に続くことはないため循環が続くことはない
        // NOTE: 見出しにオプションは使用しなくていいだろうと判断
        const serializedMDX = await serializeMDX(content);
        const MDXComponent = resolveMDXAsComponent(serializedMDX).data;

        // JSDOM へ描画し、評価後の内部のテキストだけを取得する
        const jsdom = new JSDOM(
          // NOTE: 見出しに使う見込みがないため、現状は内部で用いる components の提供はしていない
          renderToString(React.createElement(MDXComponent)),
        );
        const { textContent } = jsdom.window.document.body;

        return textContent ?? "";
      };

      const h3Regex = new RegExp("^### .*");
      if (h3Regex.test(text.trim())) {
        const headingContent = text.replace("###", "");
        const textContent = await getInnerTextContent(headingContent);
        const count = counter.count(textContent);
        // NOTE: 空白は削除
        const id = [textContent.replaceAll(" ", ""), count]
          .filter(Boolean)
          .join("-");

        return `<SubSectionTitle id="${id}">${headingContent}</SubSectionTitle>`;
      }

      const h2Regex = new RegExp("^## .*");
      if (h2Regex.test(text.trim())) {
        const headingContent = text.replace("##", "");
        const textContent = await getInnerTextContent(headingContent);
        const count = counter.count(textContent);
        // NOTE: 空白は削除
        const id = [textContent.replaceAll(" ", ""), count]
          .filter(Boolean)
          .join("-");

        return `<SectionTitle id="${id}">${headingContent}</SectionTitle>`;
      }

      return text;
    }),
  );

  return resultSplittedTexts.join("\n");
};
