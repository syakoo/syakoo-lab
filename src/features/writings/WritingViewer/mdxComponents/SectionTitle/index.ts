export { SectionTitle } from "./SectionTitle";
export { SubSectionTitle } from "./SubSectionTitle";
