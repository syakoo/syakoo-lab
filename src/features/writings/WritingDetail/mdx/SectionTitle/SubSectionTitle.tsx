"use client";

import { Link } from "@/design-system/ui";

import { subsectionTitleStyle } from "./SectionTitle.css";

type SubSectionTitleProps = {
  children: React.ReactNode;
  id?: string;
};

/**
 * SubSectionTitle コンポーネント
 */
export const SubSectionTitle: React.FC<SubSectionTitleProps> = ({
  children,
  id,
}) => {
  return (
    <h3 className={subsectionTitleStyle} id={id}>
      <Link href={`#${id}`}>{children}</Link>
    </h3>
  );
};
