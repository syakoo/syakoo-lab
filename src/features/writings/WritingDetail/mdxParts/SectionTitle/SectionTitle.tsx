import { Link } from "@/design-system/ui";

import { sectionTitleStyle } from "./SectionTitle.css";

type SectionTitleProps = {
  children: React.ReactNode;
  id?: string;
};

/**
 * SectionTitle コンポーネント
 */
export const SectionTitle: React.FC<SectionTitleProps> = ({ children, id }) => {
  return (
    <h2 className={sectionTitleStyle} id={id}>
      <Link href={`#${id}`}>{children}</Link>
    </h2>
  );
};
