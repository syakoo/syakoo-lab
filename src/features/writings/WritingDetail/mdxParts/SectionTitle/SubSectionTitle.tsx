import { Link } from "@/design-system/ui";

import { subsectionTitleStyle } from "./SectionTitle.css";
import { useSectionTitleId } from "./useSectionTitleId";

type SubSectionTitleProps = {
  children: React.ReactNode;
};

/**
 * SubSectionTitle コンポーネント
 */
export const SubSectionTitle: React.FC<SubSectionTitleProps> = ({
  children,
}) => {
  const { titleRef, sectionId } = useSectionTitleId();

  return (
    <h3 ref={titleRef} className={subsectionTitleStyle} id={sectionId}>
      <Link href={`#${sectionId}`}>{children}</Link>
    </h3>
  );
};
