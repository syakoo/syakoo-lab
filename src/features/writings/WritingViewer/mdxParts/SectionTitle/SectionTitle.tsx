import { sectionTitleStyle } from "./SectionTitle.css";
import { useSectionTitleId } from "./useSectionTitleId";
import { Link } from "@/design-system/ui";

type SectionTitleProps = {
  children: React.ReactNode;
};

/**
 * SectionTitle コンポーネント
 */
export const SectionTitle: React.FC<SectionTitleProps> = ({ children }) => {
  const { titleRef, sectionId } = useSectionTitleId();

  return (
    <h2 ref={titleRef} className={sectionTitleStyle} id={sectionId}>
      <Link href={`#${sectionId}`}>{children}</Link>
    </h2>
  );
};
