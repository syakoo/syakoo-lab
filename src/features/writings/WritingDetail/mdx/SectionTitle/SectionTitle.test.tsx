import { testStories } from "@/test/testStories";

import * as stories from "./SectionTitle.stories";

describe("SectionTitle", () => {
  testStories(stories);
});
