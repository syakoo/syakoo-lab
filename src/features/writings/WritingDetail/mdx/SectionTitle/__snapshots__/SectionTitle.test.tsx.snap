// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionTitle renders ContainElements 1`] = `
<body>
  <div>
    <h2
      class="SectionTitle_sectionTitleStyle__1gfb2v30"
    >
      <a
        class="Link_linkStyle__m4v9430"
        href="#undefined"
      >
        Sample 
        <code>
          Title
        </code>
      </a>
    </h2>
  </div>
</body>
`;

exports[`SectionTitle renders Sample 1`] = `
<body>
  <div>
    <h2
      class="SectionTitle_sectionTitleStyle__1gfb2v30"
    >
      <a
        class="Link_linkStyle__m4v9430"
        href="#undefined"
      >
        Sample Title
      </a>
    </h2>
  </div>
</body>
`;

exports[`SectionTitle renders SubSection 1`] = `
<body>
  <div>
    <h3
      class="SectionTitle_subsectionTitleStyle__1gfb2v31"
    >
      <a
        class="Link_linkStyle__m4v9430"
        href="#undefined"
      >
        Sample SubTitle
      </a>
    </h3>
  </div>
</body>
`;
